### PR TITLE
fix(model): correct Mistral4 attention and distributed MoE routing

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
## Summary

The Mistral4 checkpoint-parity test exposed four correctness bugs beyond the already-merged PP metadata fix. This PR:

- corrects Mistral4 attention scaling and long-context query scaling to match Hugging Face
- represents adaptive MoE routing bias as a replicated DTensor over each pipeline stage's DP/CP mesh
- preserves the adaptive routing bias in fp32 through mixed-precision initialization
- loads ordinary HF/adapter routing-bias tensors into the replicated DTensor checkpoint contract
- calibrates Mistral4 Phase 0/3/4 parity bounds only after ruling out checkpoint-state corruption
- installs the declared VLM media packages directly in scoped CI instead of re-resolving the complete project
- adds focused attention, distributed routing-bias, precision, and CI-config regression coverage

### 1. Mistral4 attention math differed from Hugging Face

`Mistral4MLA` inherits the DeepSeek MLA implementation. The inherited backend captured DeepSeek's YaRN-adjusted `mscale**2` softmax scale, changing the official Mistral4 scale from about `0.08839` to `0.19497` (about 2.21x). Mistral4 also applied its long-context multiplier only to the positional query slice, while Hugging Face applies it to the complete concatenated query.

The fix restores standard `qk_head_dim**-0.5` scaling and applies the multiplier to the complete query. This changes Mistral4 forward numerics in both training and inference and restores source-model parity.

### 2. Adaptive MoE routing state diverged across distributed replicas

`Gate.e_score_correction_bias` is a checkpointed routing buffer updated from observed expert-token counts. The update code already knew how to aggregate counts when the bias was a DTensor, but FSDP2 leaves registered buffers as ordinary tensors. The DTensor path therefore was not used in the common distributed configuration, and replicas updated the bias from different local batches.

This is a training correctness issue, not only a checkpoint comparison issue: different biases can change top-k expert selection, expert loads, gradients, and the resulting optimization trajectory. Consolidation retains one logical buffer value, making the divergence especially visible when comparing the warm model with a reload.

The fix explicitly distributes the buffer with `Replicate` placement over the pipeline stage's full DP/CP FSDP mesh. The existing `Partial`-to-`Replicate` update then sums expert load across all mesh dimensions before applying one identical bias update on every replica. Routing uses the bias's local replicated view, keeping ordinary local score operations free of mixed Tensor/DTensor dispatch. This removes the runtime mesh setter and manual collective path; the reduction domain now travels with the buffer itself.

Pipeline stages remain independent. `cp_size=1` is still affected when `dp_size>1`; the reproducing PP4/EP8 job had `DP=8, CP=1`.

The shared behavior change applies to models using `Gate` with `gate_bias_update_factor>0` under multi-rank FSDP2, including Mistral4, DeepSeek V3/V3.2/V4, and selected GLM and MiniMax families. Models with a zero update factor, a one-rank DP/CP mesh, or inference-only execution are unaffected.

### 3. Mistral4 mixed-precision initialization rounded adaptive routing state

AutoModel intentionally enables DeepSeek-style auxiliary-loss-free expert balancing for Mistral4 fine-tuning. The published Hugging Face and native Mistral4 checkpoints do not contain this bias, so the state-dict adapter injects an fp32 zero buffer. Zero preserves the released model's initial routing; AutoModel then adjusts the bias by `1e-3` after each optimizer step to discourage expert overload.

The text and multimodal Mistral4 initializers used a raw whole-model `to(bfloat16)`, which also cast the registered correction-bias buffers to bf16. This violated the shared MoE precision contract: small rounding differences can accumulate in the bias and flip discrete top-k expert selection.

Both wrappers now declare `e_score_correction_bias` as strict fp32 state and use the shared precision-aware model cast. The state remains a persistent registered buffer—not a parameter—and is preserved in fp32 before and after FSDP2/DTensor parallelization. Focused CPU tests exercise both production initialization paths.

### 4. Base-checkpoint loading mixed ordinary tensors with the replicated DTensor

Once the persistent bias became a replicated DTensor, the checkpoint-robustness setup exposed a second state contract. HF adapters produce ordinary full tensors—for Mistral4, the adapter injects an fp32 zero bias because the released checkpoint omits it—while a model that has already been FSDP-parallelized owns a replicated DTensor buffer. PyTorch cannot execute `copy_` with one ordinary tensor and one DTensor, so the base checkpoint failed to load before training or parity evaluation began.

The shared `Gate` loader now converts an ordinary full bias to a replicated DTensor on the live buffer's mesh before PyTorch copies state. It preserves the live fp32 dtype and device, accepts native DCP DTensors without conversion, and fails explicitly for a shape mismatch or any non-replicated placement. Keeping this at the state-owning `Gate` boundary gives Mistral4, ERNIE, MiniMax, HY, and other adapters one checkpoint contract instead of duplicating DTensor knowledge across model-specific adapters.

## Checkpoint-parity conclusion

The remaining Phase 3 delta is not consolidation corruption:

- repeated forwards within each runtime were bit-identical
- two 32-rank fingerprints found identical tensor counts, no missing/extra state, and zero sampled parameter or buffer mismatches
- a fresh native-DCP load and a fresh consolidated load had the same `0.01264910` KL relative to the warm Phase 2 runtime
- the fresh DCP and consolidated logits were bit-identical to each other (`0.00000000` max KL)

Phase 3 therefore bounds the warm-versus-fresh distributed BF16 runtime boundary, not a state-format difference. Post-fix Phase 3 ranged from about `0.0033` to `0.0378`; `kl_threshold: 0.05` provides limited headroom. Phase 4 compares native TE/SDPA/EP execution with vanilla Hugging Face and ranged from about `0.0214` to `0.0424`, also bounded at `0.05`. Phase 0 retains tighter source-parity bounds: max KL `0.01`, mean KL `0.002`, and cosine `0.999`.

Resume-loss behavior remains separately tracked in [AMINT-227](https://linear.app/nvidia/issue/AMINT-227/redesign-checkpoint-resume-robustness-around-a-shared-training).

## CI change

The scoped VLM launcher now installs the already-declared media packages directly instead of re-resolving the complete AutoModel project and an unrelated inaccessible Git dependency. This changes test setup only, not package declarations or runtime model behavior.

## Validation

- Mistral4 unit tests: `60 passed, 70 skipped`
- Mistral4 plus common dtype tests: `104 passed, 70 skipped`
- MoE gate/parallelizer tests on the DTensor head: `171 passed, 3 skipped`
- shared Gate plus Mistral4/ERNIE/MiniMax/HY adapter regression set: `212 passed, 3 skipped`
- real two-rank Gloo checkpoint test covers ordinary fp32 tensor, native DCP DTensor, and rejection of a sharded target: `1 passed`
- CI resolver tests: `40 passed`
- targeted Ruff format/check and `git diff --check`: passed
- one-rank DTensor checkpoint probe round-tripped the replicated buffer exactly
- one-rank meta-device probe preserved DTensor placement through production-style materialization
- the two-rank ERNIE FSDP regression verifies `Replicate` placement, local routing, and a numerically correct globally aggregated bias update
- [Mistral4 pipeline 60880218](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60880218): the EOS retry reached model setup and exposed the ordinary-Tensor-to-DTensor base-load mismatch before training
- [current Mistral4 pipeline 60960166](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60960166): exact `mistral4_medpix` checkpoint-robustness rerun on the shared loader fix
- [cross-model MoE pipeline 60960168](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60960168): exact ERNIE 4.5 checkpoint-robustness and MiniMax M2.7 adaptive-bias training controls
- [gate-bias diagnostic pipeline 60587316](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60587316): confirmed identical routing bias after DP/CP synchronization
- [state-fingerprint pipeline 60597238](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60597238): all 32 PP4/EP8 ranks had matching state fingerprints; Phase 3 measured `0.0032946` and resume passed
- [fresh-DCP control pipeline 60600120](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60600120): native DCP and consolidated reloads produced bit-identical logits; Phase 4 and resume passed
- [final production pipeline 60606401](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/60606401): completed all 50 PP4/EP8 training steps, Phase 0/3/4, and resume with no masked-scatter assertion

The static PP media-cursor, checkpoint-resume metadata, and timeout fixes were merged separately in [#3344](https://github.com/NVIDIA-NeMo/Automodel/pull/3344) and are not part of this PR's effective diff.

Related: AMINT-230, AMINT-227, AM-715

### CODE REVIEW

**Comment by yuhezhang-ai** [comment]
/ok to test [31a4d6b](https://github.com/NVIDIA-NeMo/Automodel/pull/3348/commits/31a4d6b03f9ad05d9e3a86cf20b6d768a043d816)

**Comment by yuhezhang-ai** [comment]
/ok to test [2a04739](https://github.com/NVIDIA-NeMo/Automodel/pull/3348/commits/2a04739647f4733249a2f04868373fe66eaaddb3)

File: nemo_automodel/components/moe/parallelizer.py [hunk 1]
```diff
@@ -682,6 +683,12 @@ def apply_fsdp(
 
     for block in _iter_moe_blocks(model, _model):
         moe_module = _get_moe_module(block)
+        gate = getattr(moe_module, "gate", None)
+        if isinstance(gate, Gate):
+            # FSDP2 does not convert buffers to DTensors. Give the learned
+            # load-balancing bias the same DP/CP mesh used for dense state so
+            # every replica applies one globally consistent routing update.
+            gate._set_expert_load_mesh(fsdp_mesh)
```
Comment (akoumpa) [inline nemo_automodel/components/moe/parallelizer.py]: @yuhezhang-ai , I'm wondering if instead of _set_expert_load_mesh, we could make e_score_correction_bias a DTensor and leverage the existing code -- although I'm not 100% sure of the existing code's correctness under DPxCP. TBH, I'm looking to avoid a setter for this case.

**Comment by yuhezhang-ai** [comment]
/ok to test [8985508](https://github.com/NVIDIA-NeMo/Automodel/pull/3348/commits/898550876a38687f52699a382d12c23ac8081980)

**Comment by yuhezhang-ai** [comment]
/ok to test [4146c10](https://github.com/NVIDIA-NeMo/Automodel/pull/3348/commits/4146c10d08a69e591b16f3a038f0b40280066671)

**Comment by yuhezhang-ai** [comment]
/ok to test [ada0bd0](https://github.com/NVIDIA-NeMo/Automodel/pull/3348/commits/ada0bd070efd950ed55fdf14e968684a9215ed40)

File: nemo_automodel/components/moe/parallelizer.py [hunk 2]
```diff
@@ -682,6 +683,16 @@ def apply_fsdp(
 
     for block in _iter_moe_blocks(model, _model):
         moe_module = _get_moe_module(block)
+        gate = getattr(moe_module, "gate", None)
+        if isinstance(gate, Gate) and gate.e_score_correction_bias is not None:
```
Comment (akoumpa) [inline nemo_automodel/components/moe/parallelizer.py:687]: thank you @yuhezhang-ai

**Review by akoumpa — APPROVED** [review]
